msal/application.py

```diff
@@ -1398,7 +1399,10 @@ def _acquire_token_silent_from_cache_and_possibly_refresh_it(
             if account and account.get("authority_type") == _AUTHORITY_TYPE_CLOUDSHELL:
                 return self._acquire_token_by_cloud_shell(scopes, data=data)
 
-            if self._enable_broker and account is not None:
+            if self._enable_broker and account and account.get("account_source") in (
+                _GRANT_TYPE_BROKER,  # Broker successfully established this account previously.
+                None,  # Unknown data from older MSAL. Broker might still work.
+            ):
```

Comment:
Besides fixing device code flow (https://github.com/AzureAD/microsoft-authentication-library-for-python/issues/563), another great benefit of this PR is that enabling WAM will not break the current login context established by auth code flow. I provided more details in https://github.com/Azure/azure-cli/pull/28085#issuecomment-1870093562.

Reply:
Yes, that was an intentional design and precise implementation. We did not "advertise" it, though, because such an effect is controversial. Some might perceive it as lack of enforcement to utilize broker.

Luckily, that debate is not really necessary in Azure CLI's context, because [Azure CLI's broker instruction suggests end user to clear local cache before enabling broker](https://learn.microsoft.com/en-us/cli/azure/authenticate-azure-cli-web-account-manager).

Let's end this topic here. Further question can be carried out offline.